File: src/main/java/com/raiderrobotix/_2020/OperatorInterface.java
package com.raiderrobotix._2020;

import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.Sendable;
import edu.wpi.first.wpilibj.buttons.JoystickButton;
import edu.wpi.first.wpilibj.smartdashboard.SendableBuilder;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

public class OperatorInterface implements Sendable {

	public static final OperatorInterface input = new OperatorInterface();

	private static final int LEFT_JOYSTICK_PORT = 0, RIGHT_JOYSTICK_PORT = 1, OPERATOR_JOYSTICK_PORT = 2;

	private static final double JOYSTICK_DEADBAND = 0.15;

	// Joysticks
	private static final Joystick leftStick = new Joystick(LEFT_JOYSTICK_PORT);
	private static final Joystick rightStick = new Joystick(RIGHT_JOYSTICK_PORT);
	private static final Joystick operatorStick = new Joystick(OPERATOR_JOYSTICK_PORT);

	private static final JoystickButton[] operator = new JoystickButton[12];
	private static final JoystickButton[] left = new JoystickButton[12];
	private static final JoystickButton[] right = new JoystickButton[12];

	static {
		for (int button = 0; button < 12; button++) {
			left[button] = new JoystickButton(leftStick, button);
			right[button] = new JoystickButton(rightStick, button);
			operator[button] = new JoystickButton(operatorStick, button);
		}
	}

	private OperatorInterface() {
		SmartDashboard.putData(this);
	}

	public double getLeftY() {
		double ret = leftStick.getY();
		return Math.abs(ret) > JOYSTICK_DEADBAND ? ret : 0.0;
	}

	public double getRightY() {
		double ret = rightStick.getY();
		return Math.abs(ret) > JOYSTICK_DEADBAND ? ret : 0.0;
	}

	/**
	 * Down on Joystick is positive, up is negative
	 */
	public double getOperatorY() {
		return operatorStick.getY();
	}

	public boolean getLeftButton(int button) {
		return leftStick.getRawButton(button);
	}

	public boolean getRightButton(int button) {
		return rightStick.getRawButton(button);
	}

	public boolean getOperatorButton(int button) {
		return operatorStick.getRawButton(button);
	}

	public boolean getOperatorTrigger() {
		return operatorStick.getTrigger();
	}

	@Override
	public void initSendable(SendableBuilder builder) {
		builder.addDoubleProperty("lY", this::getLeftY, null);
		builder.addDoubleProperty("rY", this::getRightY, null);
		builder.addDoubleProperty("oY", this::getOperatorY, null);
	}
}

File: src/main/kotlin/com/raiderrobotix/_2020/OperatorInterface.kt
package com.raiderrobotix._2020

import com.raiderrobotix._2020.commands.colorwheel.positionControl
import com.raiderrobotix._2020.subsystems.ColorWheel
import com.raiderrobotix._2020.subsystems.ColorWheel.zeroOutColor
import com.raiderrobotix._2020.subsystems.Intake
import com.raiderrobotix._2020.subsystems.Shooter
import com.raiderrobotix._2020.subsystems.Trolley
import edu.wpi.first.wpilibj.GenericHID
import edu.wpi.first.wpilibj.Joystick
import org.ghrobotics.lib.wrappers.hid.FalconHID
import org.team2471.frc.lib.input.whenTrue
import org.team2471.frc.lib.input.whileTrue
import kotlin.math.abs

object OperatorInterface {
	
	private const val LEFT_JOYSTICK_PORT = 0
	private const val RIGHT_JOYSTICK_PORT = 1
	private const val OPERATOR_JOYSTICK_PORT = 2
	private const val JOYSTICK_DEADBAND = 0.15
	
	// Joysticks
	
	internal val left = Joystick(LEFT_JOYSTICK_PORT)
	internal val right = Joystick(RIGHT_JOYSTICK_PORT)
	internal val operator = Joystick(OPERATOR_JOYSTICK_PORT)
	
	
	init {
		val flip = 2
		
		val shooter = 1
		({ operator[shooter] }).whenTrue { Shooter.speed = 1.0 }
		({ !operator[shooter] }).whenTrue { Shooter.reset() }
		
		val conveyer = 11
		({ operator[conveyer] && !operator[flip] }).whenTrue { Intake.speed = 1.0 }
		({ operator[conveyer] && operator[flip] }).whenTrue { Intake.speed = -0.5 }
		({ !operator[conveyer] }).whenTrue { Intake.reset() }
		
		val roller = 12
		({ operator[roller] && !operator[flip] }).whenTrue {
			Intake.outer.speed = 0.6
		}
		({ operator[roller] && operator[flip] }).whenTrue {
			Intake.outer.speed = -0.7
		}
		({ !operator[roller] }).whenTrue { Intake.reset() }
		
		val trolley = 7
		({ operator[trolley] && !operator[flip] }).whenTrue { Trolley.speed = 1.0 }
		({ operator[trolley] && operator[flip] }).whenTrue { Trolley.speed = -1.0 }
		({ !operator[trolley] }).whenTrue { Trolley.reset() }
		
		//Cowl
		val cowl = 5
		({ operator[cowl] }).whenTrue { Shooter.cowlSpeed = 0.5 }
		({ operator[3] }).whenTrue { Shooter.cowlSpeed = -0.5 }
		({ !operator[cowl] && !operator[3] }).whenTrue { Shooter.cowlSpeed = 0.0 }
		
		//Turn to Color
		val colorWheel = 10
		({ right[colorWheel] }).whileTrue {
			positionControl()
		}
		({ !right[colorWheel] }).whenTrue { ColorWheel.reset() }
		
		({ right[11] }).whenTrue { zeroOutColor(iter = 20) }
		
		
	}
	
	/**
	 * Down on Joystick is positive, up is negative
	 */
	val leftY: Double
		get() {
			val ret = left.y
			return if (abs(ret) > JOYSTICK_DEADBAND) ret else 0.0
		}
	
	val rightY: Double
		get() = right.y.takeIf { abs(it)  > JOYSTICK_DEADBAND } ?: 0.0
	
	val operatorY: Double
		get() = operator.y.takeIf { abs(it) > JOYSTICK_DEADBAND }
			?: 0.0
	
	val operatorTrigger: Boolean
		get() = operator.trigger
	
	operator fun Joystick.get(button: Int) = getRawButton(button)

	operator fun <T : GenericHID> FalconHID<T>.get(button: Int) = getRawButton(button)
	
}
